src/mbgl/renderer/source_state.cpp

```diff
@@ -0,0 +1,157 @@
+#include <mbgl/renderer/source_state.hpp>
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/renderer/render_tile.hpp>
+#include <mbgl/util/logging.hpp>
+
+namespace mbgl {
+
+using namespace mbgl::style::conversion;
+
+void SourceFeatureState::updateState(const optional<std::string>& sourceLayerID, const std::string& featureID, const Convertible& newState) {
+
+    if (!isObject(newState)) {
+        Log::Error(Event::General, "Feature state must be an object");
+        return;
+    }
+
+    std::string sourceLayer = sourceLayerID ? *sourceLayerID : "";
+
+    std::string stateKey;
+    Value stateValue;
+
+    const std::function<optional<Error> (const std::string&, const Convertible&)> convertFn = [&] (const std::string& k, const Convertible& v) -> optional<Error> {
+        optional<Value> value = toValue(v);
+        if (value) {
+            stateValue = *value;
+        } else {
+            if (isArray(v)) {
+                std::vector<Value> array;
+                size_t length = arrayLength(v);
+                for (size_t i = 0; i < length; i++) {
+                    optional<Value> arrayVal = toValue(arrayMember(v, i));
+                    if (arrayVal) {
+                        array.emplace_back(*arrayVal);
+                    }
+                }
+                std::unordered_map<std::string, Value> result;
+                result[k]= array;
+                stateValue = result;
+                return {};
+
+            } else if (isObject(v)) {
+                eachMember(v, convertFn);
+            }
+        }
+        stateKey = k;
+        PropertyMap stateResult = {{ stateKey, stateValue }};
+        stateChanges[sourceLayer][featureID][stateKey] = stateValue;
+        return {};
+    };
+
+    eachMember(newState, convertFn);
+}
+
+PropertyMap SourceFeatureState::getState(const optional<std::string>& sourceLayerID, const std::string& featureId) const {
+    std::string sourceLayer = sourceLayerID ? *sourceLayerID : "";
+    PropertyMap base, changes;
+    auto states = state.find(sourceLayer);
+    if (states != state.end()) {
+        auto featureStateBase = states->second.find(featureId);
```

Comment:
could you add comment to the header for featureStateBase, so it would be easier to understand what it contains.